mvrx/src/main/kotlin/com/airbnb/mvrx/PersistState.kt

```diff
@@ -73,49 +71,75 @@ private fun assertCollectionPersistability(value: Any?) {
-    if (item !is Serializable && item !is Parcelable) throw IllegalStateException("Cannot parcel ${item::class.simpleName}")
+    if (item !is Serializable && item !is Parcelable) throw IllegalStateException("Cannot parcel ${item::class.java.simpleName}")
 }
 
 private fun <T : Any?> Bundle.putAny(key: String?, value: T): Bundle {
     when (value) {
         is Parcelable -> putParcelable(key, value)
         is Serializable -> putSerializable(key, value)
         null -> putString(key, null)
-        else -> throw IllegalStateException("Cannot persist $key. It must be null, Serializable, or Parcelable.")
+        else -> error("Cannot persist $key. It must be null, Serializable, or Parcelable.")
     }
     return this
 }
 
 /**
  * Updates the initial state object given state persisted with [PersistState] in a [Bundle].
  */
-internal fun <T : Any> Bundle.restorePersistedState(initialState: T): T {
+internal fun <T : MvRxState> Bundle.restorePersistedState(initialState: T, validation: Boolean = false): T {
+    val jvmClass = initialState::class.java
+    val constructor = jvmClass.constructors.firstOrNull { it.parameterAnnotations.any { it.any { it is PersistState } } } ?: return initialState
+
     // If we don't set the correct class loader, when the bundle is restored in a new process, it will have the system class loader which
-    // can't unmarshall any custom classes.
-    val stateClass = initialState::class
-    classLoader = stateClass.java.classLoader
-    val constructor = stateClass.primaryConstructor ?: throw IllegalStateException("${stateClass.simpleName} has no primary constructor!")
-    val persistedConstructorParamNames = constructor.parameters.asSequence()
-        .filter { it.name != null }
-        .filter { it.annotations.any { it.annotationClass == PersistState::class } }
-        .map { it.name }
-        .toSet()
-    if (persistedConstructorParamNames.isEmpty()) {
-        return initialState
-    }
+    // can't unmarshal any custom classes.
+    classLoader = jvmClass.classLoader
+
+    // For data classes, Kotlin generates a static functionc called copy$default.
+    // The first parameter is the object to copy from.
+    // The next parameters are all of parameters to copy (it's jvm bytecode/java so there are no optional parameters in the generated method).
+    // The next parameter(s) are a bitmask. Each parameter index corresponds to one bit in the int.
+    //     If the bitmask is 1 for a given parameter then the new object will have the original object's value and the parameter value will be ignored.
+    //     If the bitmask is 0 then it will use the value from the parameter.
+    //     There is 1 bitmask for every 32 parameters. If there are 48 parameters, there will be 2 bitmasks. Parameter 33 will be the first bit of the 2nd bitmask.
+    // The last parameter is ignored. It can be null.
+    val copyFunction = jvmClass.declaredMethods.first { it.name == "copy\$default" }
```

Comment:
Probably worth adding a proguard rule for this. Combine it with the persisted state base class

```
-if class * extends path.to.MvRxState
-keepclassmembers @path.to.PersistedState class <1> {
    synthetic <1> copy$default(...);
}
```

Reply:
@gpeal @ZacSweers

The old rule would have worked but it is not optimal and I also had updated it for this change already. 
I retested and updated again to:
```
# The constructor as well as the copy$default() method and the component*() methods of the Kotlin data classes used as
# the state in MvRx are read via reflection which cause trouble with Proguard if they are not kept.
-keepclassmembers class ** implements com.airbnb.mvrx.MvRxState {
   public <init>(...);
   synthetic *** copy$default(...);
   public *** component*();
}
```
this should take care of all of those.